From:	Matthew Garrett <mjg@redhat.com>
Date:	Thu, 20 Sep 2012 10:40:58 -0400
Subject: [PATCH V2 03/10] x86: Lock down IO port access in secure boot environments
Patch-mainline: Not yet, reviewing
References: fate#314486
Target: SLE-11 SP3

IO port access would permit users to gain access to PCI configuration
registers, which in turn (on a lot of hardware) give access to MMIO register
space. This would potentially permit root to trigger arbitrary DMA, so lock
it down by default.

Signed-off-by: Matthew Garrett <mjg@redhat.com>
Acked-by: Lee, Chun-Yi <jlee@suse.com>
Automatically created from "patches.suse/0003_x86:_Lock_down_IO_port_access_in_secure_boot_environments_v2.patch" by xen-port-patches.py

--- sle11sp3.orig/arch/x86/kernel/ioport-xen.c	2011-04-12 16:53:32.000000000 +0200
+++ sle11sp3/arch/x86/kernel/ioport-xen.c	2012-12-04 13:55:02.000000000 +0100
@@ -28,7 +28,7 @@ asmlinkage long sys_ioperm(unsigned long
 
 	if ((from + num <= from) || (from + num > IO_BITMAP_BITS))
 		return -EINVAL;
-	if (turn_on && !capable(CAP_SYS_RAWIO))
+	if (turn_on && (!capable(CAP_SYS_RAWIO) || !capable(CAP_COMPROMISE_KERNEL)))
 		return -EPERM;
 
 	/*
@@ -74,7 +74,7 @@ long sys_iopl(unsigned int level, struct
 		return -EINVAL;
 	/* Trying to gain more privileges? */
 	if (level > old) {
-		if (!capable(CAP_SYS_RAWIO))
+		if (!capable(CAP_SYS_RAWIO) || !capable(CAP_COMPROMISE_KERNEL))
 			return -EPERM;
 	}
 	t->iopl = level << 12;
